src/golang/lib/workflow/artifact/migrate.go

```diff
@@ -0,0 +1,101 @@
+package artifact
+
+import (
+	"context"
+
+	"github.com/aqueducthq/aqueduct/lib/collections/artifact"
+	"github.com/aqueducthq/aqueduct/lib/collections/artifact_result"
+	"github.com/aqueducthq/aqueduct/lib/collections/shared"
+	"github.com/aqueducthq/aqueduct/lib/collections/workflow_dag"
+	"github.com/aqueducthq/aqueduct/lib/database"
+	"github.com/aqueducthq/aqueduct/lib/storage"
+	"github.com/aqueducthq/aqueduct/lib/workflow/utils"
+	"github.com/sirupsen/logrus"
+)
+
+// Migrate moves from workflow artifact result content from `oldConf` to `newConf`.
+func Migrate(
+	ctx context.Context,
+	oldConf *shared.StorageConfig,
+	newConf *shared.StorageConfig,
+	dagReader workflow_dag.Reader,
+	dagWriter workflow_dag.Writer,
+	artifactReader artifact.Reader,
+	artifactResultReader artifact_result.Reader,
+	db database.Database,
+) error {
+	// Wait until there are no more workflow runs in progress
+	lock := utils.NewExecutionLock()
+	if err := lock.Lock(); err != nil {
+		return err
+	}
+	defer func() {
+		unlockErr := lock.Unlock()
+		if unlockErr != nil {
+			logrus.Errorf("Unexpected error when unlocking workflow execution lock: %v", unlockErr)
+		}
+	}()
+
+	oldStore := storage.NewStorage(oldConf)
+	newStore := storage.NewStorage(newConf)
+
+	txn, err := db.BeginTx(ctx)
+	if err != nil {
+		return err
+	}
+	defer database.TxnRollbackIgnoreErr(ctx, txn)
+
+	dags, err := dagReader.GetWorkflowDags(ctx, nil, txn)
+	if err != nil {
+		return err
+	}
+
+	for _, dag := range dags {
+		if dag.EngineConfig.Type == shared.AirflowEngineType {
+			// We cannot migrate content for Airflow workflows
+			continue
+		}
+
+		artifacts, err := artifactReader.GetArtifactsByWorkflowDagId(ctx, dag.Id, txn)
+		if err != nil {
+			return err
+		}
+
+		for _, artifact := range artifacts {
+			artifactResults, err := artifactResultReader.GetArtifactResultsByArtifactId(ctx, artifact.Id, txn)
+			if err != nil {
+				return err
+			}
+
+			// For each artifact result, move the content from `oldStore` to `newStore`
+			for _, artifactResult := range artifactResults {
+				val, err := oldStore.Get(ctx, artifactResult.ContentPath)
```

Comment:
A "migration" makes it sound like we should be deleting the data from the old store, once we've successfully written to the new store + updated the dag. Is this something we want to do, or are we ok leaving orphaned copies of data in the old store - I'm worried it'll cause confusion somehow.

Reply:
@vsreekanti Are we ok with orphaned data or do we want to clean this up explicitly. I would rather not delete as part of the migration just in case something goes wrong. We wouldn't want to lose customer data.

Reply:
I'd prefer not to leave stuff lying around that's then orphaned because we'll lose the context for it, no? Maybe we can do it in two stages: Move everything first with a verification step, then delete?

Reply:
Added as part of #561